scripts/check_ecosystem.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+"""Check two versions of ruff against a corpus of open-source code.
+
+Example usage:
+
+    scripts/check_ecosystem.py <path/to/ruff1> <path/to/ruff2>
+"""
+
+# ruff: noqa: T201
+
+import argparse
+import asyncio
+import difflib
+import heapq
+import re
+import tempfile
+from asyncio.subprocess import PIPE, create_subprocess_exec
+from contextlib import asynccontextmanager
+from pathlib import Path
+from typing import TYPE_CHECKING, NamedTuple, Self
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator, Iterator, Sequence
+
+
+class Repository(NamedTuple):
+    """A GitHub repository at a specific ref."""
+
+    org: str
+    repo: str
+    ref: str
+
+    @asynccontextmanager
+    async def clone(self: Self) -> "AsyncIterator[Path]":
+        """Shallow clone this repository to a temporary directory."""
+        with tempfile.TemporaryDirectory() as tmpdir:
+            process = await create_subprocess_exec(
+                "git",
+                "clone",
+                "--config",
+                "advice.detachedHead=false",
+                "--quiet",
+                "--depth",
+                "1",
+                "--no-tags",
+                "--branch",
+                self.ref,
+                f"https://github.com/{self.org}/{self.repo}",
+                tmpdir,
+            )
+
+            await process.wait()
+
+            yield Path(tmpdir)
+
+
+REPOSITORIES = {
+    "zulip": Repository("zulip", "zulip", "main"),
+    "bokeh": Repository("bokeh", "bokeh", "branch-3.2"),
+}
+
+SUMMARY_LINE_RE = re.compile(r"^Found \d+ error.*$")
+
+
+class RuffError(Exception):
+    """An error reported by ruff."""
+
+
+async def check(*, ruff: Path, path: Path) -> "Sequence[str]":
+    """Run the given ruff binary against the specified path."""
+    proc = await create_subprocess_exec(
+        ruff.absolute(),
+        "check",
+        "--no-cache",
+        "--exit-zero",
+        ".",
```

Comment:
Running ruff in the isolated mode with all rules enabled could help a bit with 

> However, the most valuable way of setting this up is actually to test messy codebases that fail many checks, not ones that already use ruff, because that would be the only way to catch PRs introducing false negatives.

Reply:
If we're looking to catch false negatives, we could consider running over CPython? It's a fairly diverse codebase.

Reply:
CPython could work but it includes a lot of non-Python code, meaning it would do a bunch of extra work to download/clone files that it won't end up checking. I do like the idea of using isolated mode and setting all rules to enabled so I've made that change.

Reply:
The downside of using isolated mode is that in some cases, it will _avoid_ exercising certain behaviors. For example, Airflow uses `namespace-packages`, and if we run in isolated mode, we might miss breakages to our namespace packages handling, since in isolated mode, it's effectively running with `namespace-packages = []`. I don't know how to resolve that though. I suppose we could run without `--isolated`, but with `--select ALL`?

Reply:
I may end up tweaking this.